src/dataflow/source/file.rs

```diff
@@ -213,80 +280,128 @@ where
-            let next_time = if cap_time > sys_time {
-                if cap_time != sys_time + 1 {
-                    warn!(
-                        "{}: fast-forwarding out-of-order Unix timestamp {}ms ({} -> {})",
-                        n2,
-                        cap_time - sys_time,
-                        sys_time,
-                        cap_time,
-                    );
-                }
-                next_activation_duration = Duration::from_millis(cap_time - sys_time);
-                cap_time
-            } else {
-                cap.downgrade(&sys_time);
-                sys_time + 1
-            };
-
+            // Number of records read for this particular activation
             let mut records_read = 0;
 
-            let mut session = output.session(cap);
-            while records_read < MAX_RECORDS_PER_INVOCATION {
-                match rx.try_recv() {
-                    Ok(record) => {
-                        records_read += 1;
-                        total_records_read += 1;
-                        session.give((record, Some(total_records_read)));
+            if read_file {
+                // Check if the capability can be downgraded (this is independent of whether
+                // there are new messages that can be processed) as timestamps can become
+                // closed in the absence of messages
+                downgrade_capability(
+                    &id,
+                    cap,
+                    &mut last_processed_offset,
+                    &mut last_closed_ts,
+                    &timestamp_histories,
+                );
+
+                // Check if there was a message buffered and if we can now process it
+                // If we can now process it, clear the buffer and proceed to poll from
+                // consumer. Else, exit the function
+                buffer = if let Some(message) = buffer.take() {
+                    Some(message)
+                } else {
+                    match rx.try_recv() {
+                        Ok(record) => {
+                            records_read += 1;
+                            current_msg_offset += 1;
+                            Some(record)
+                        }
+                        Err(TryRecvError::Empty) => None,
+                        Err(TryRecvError::Disconnected) => {
+                            return SourceStatus::Done;
+                        }
+                    }
+                };
+
+                while let Some(message) = buffer.take() {
+                    let ts = find_matching_timestamp(&id, current_msg_offset, &timestamp_histories);
+                    match ts {
+                        None => {
+                            // We have not yet decided on a timestamp for this message,
+                            // we need to buffer the message
+                            buffer = Some(message);
+                            activator.activate();
```

Comment:
we discussed this on #timely-differential on slack. Conclusion was that it's fine because calling `activate` over and over will not block other operators from running.